cupertino/src/commonMain/kotlin/io/github/alexzhirkevich/cupertino/CupertinoSwipeBox.kt

```diff
@@ -1,1043 +1,240 @@
-    animationSpec : FiniteAnimationSpec<Float> = CupertinoSwipeBoxDefaults.AnimationSpec,
-    confirmValueChange: (CupertinoSwipeBoxValue) -> Boolean = { true },
-) : CupertinoSwipeBoxState {
-
-    val density = LocalDensity.current
-
-    val state = rememberSaveable(
-        saver = CupertinoSwipeBoxState.Saver(
-            dismissThreshold = dismissThreshold,
-            animationSpec = animationSpec,
-            confirmValueChange = confirmValueChange,
-            density = density,
-        )
-    ) {
-        CupertinoSwipeBoxState(
-            density = density,
-            initialValue = initialValue,
-            animationSpec = animationSpec,
-            dismissThreshold = dismissThreshold,
-            confirmValueChange = confirmValueChange,
-        )
-    }
-
-    if (sharedController != null) {
-        LaunchedEffect(state) {
-            launch {
-                sharedController.expanded
-                    .filter { it !== state }
-                    .collectLatest {
-                        state.animateTo(CupertinoSwipeBoxValue.Collapsed)
-                    }
-            }
-            launch {
-                snapshotFlow {
-                    state.offset != 0f && !state.offset.isNaN()
-                }.filter { it }.collectLatest {
-                    sharedController.expanded.emit(state)
-                }
-            }
-        }
-    }
-
-    if (scrollableState != null) {
-        LaunchedEffect(scrollableState) {
-            snapshotFlow {
-                scrollableState.isScrollInProgress
-            }.filter { it }.collectLatest {
-                if (state.dismissDirection != CupertinoSwipeBoxValue.Collapsed &&
-                    state.confirmValueChange(CupertinoSwipeBoxValue.Collapsed)
-                ) {
-                    state.animateTo(CupertinoSwipeBoxValue.Collapsed)
-                }
-            }
-        }
-    }
-
-    return state
-}
-
-enum class SwipeBoxBehavior {
-
-    /**
-     * Swipe in this direction is completely disabled
-     * */
-    Disabled,
-
-    /**
-     * Background items can be expanded and settled but can't be dismissed
-     * */
-    Expandable,
-
-    /**
-     * Background items can be expanded and the first item can be swiped out to the end hiding other items
-     * */
-    Dismissible
+fun CupertinoSwipeBox(
+    state: AnchoredDraggableState<SwipeBoxStates> = rememberSimpleCupertinoSwipeBoxState(),
+    modifier: Modifier = Modifier,
+    actionItemWidth: Dp = SimpleCupertinoSwipeBoxDefaults.actionItemWidth,
+    height: Dp = SimpleCupertinoSwipeBoxDefaults.actionItemHeight,
+    fullExpansionStart: Boolean = SimpleCupertinoSwipeBoxDefaults.allowFullSwipe,
+    fullExpansionEnd: Boolean = SimpleCupertinoSwipeBoxDefaults.allowFullSwipe,
+    startActionItem: Pair<(@Composable RowScope.() -> Unit), (() -> Unit)>? = null,
+    endActionItem: Pair<(@Composable RowScope.() -> Unit), (() -> Unit)>? = null,
```

Comment:
Yeah I agree, but I haven't been able to think of a better method yet.

I think maybe we refactor it so users pass in a refactored SwipeBoxBoxItem that has its own onClick that we can reference. 

I do think users should only have to pass in one object for start and one for end actions, as opposed to a separate parameter for the full swipe action.

Reply:
The SwipeBox public API should remain unchanged. It is the only appropreate variant. At most, actions lambda can be split to 2 separate composable lambdas.

Reply:
Again, if it can't be done, better remove the SwipeBox for now and fix it later

Reply:
Just making sure we're clear, you're suggesting that the current API - 

```
CupertinoSwipeBox(
        modifier = Modifier
            .fillMaxWidth(),
        state = state,
        startToEndBehavior = SwipeBoxBehavior.Expandable,
        items = {
            when {
                state.dismissDirection.isTowardsStart -> {
                    CupertinoSwipeBoxItem(
                        onClick = {
                            scope.launch {
                                if (state.currentValue == CupertinoSwipeBoxValue.DismissedToStart) {
                                    state.reset()
                                } else {
                                    state.animateTo(CupertinoSwipeBoxValue.DismissedToStart)
                                }
                            }
                        },
                        color = CupertinoColors.systemRed,
                        icon = {
                            Icon(
                                imageVector = CupertinoIcons.Filled.Trash,
                                contentDescription = "Delete"
                            )
                        },
                        label = {
                            Text("Delete")
                        }
                    )
                    CupertinoSwipeBoxItem(
                        onClick = {
                            scope.launch { state.reset() }
                        },
                        color = CupertinoColors.systemOrange,
                        icon = {
                            Icon(
                                imageVector = CupertinoIcons.Filled.SpeakerSlash,
                                contentDescription = "Mute"
                            )
                        },
                        label = {
                            Text("Mute")
                        }
                    )
                }

                state.dismissDirection.isTowardsEnd -> {
                    CupertinoSwipeBoxItem(
                        onClick = {
                            scope.launch { state.reset() }
                        },
                        color = CupertinoColors.systemGray,
                        icon = {
                            Icon(
                                imageVector = CupertinoIcons.Filled.BubbleLeft,
                                contentDescription = "Unread"
                            )
                        },
                        label = {
                            Text("Unread")
                        }
                    )
                    CupertinoSwipeBoxItem(
                        onClick = {
                            scope.launch { state.reset() }
                        },
                        color = CupertinoColors.systemGreen,
                        icon = {
                            Icon(
                                imageVector = CupertinoIcons.Filled.Pin,
                                contentDescription = "Pin"
                            )
                        },
                        label = {
                            Text("Pin")

                        }
                    )
                }

                else -> {
                    // Empty content on collapsed state to avoid clipping artifacts
                }
            }
        }
    ) {
        Box(
            modifier = Modifier
                .fillMaxSize()
                .height(72.dp)
                .background(LocalContainerColor.current)
                .clickable {}
                .padding(
                    start = CupertinoSectionDefaults.PaddingValues
                        .calculateStartPadding(LocalLayoutDirection.current)
                )
        ) {
            Text(
                modifier = Modifier
                    .align(Alignment.CenterStart),
                text = "Swipe horizontally"
            )
        }
    }
```

is **the only appropriate variant**? Why should users need to specify the items per Swipe direction while also supplying the swipe behavior. I mean there's an empty else statement added to avoid clipping artifacts?

Also, the SwipeBox is marked as experimental, that's what we're doing with the API.

> Besides, pairs are unstable, and this code will recompose on every single pixel of swipe gesture without strong skipping mode enabled

I can remove the pairs, but I can't say I have a 100% grasp of everything you tried to do in the original ~1600 line of code [CupertinoSwipeBox](https://github.com/alexzhirkevich/compose-cupertino/blob/master/example/shared/src/commonMain/kotlin/cupertino/CupertinoWidgetsScreen.kt).

Overall, I think the eventual goal of `.anchoredDraggable` will be similar to iOS's `.swipeEdge`:
- https://developer.apple.com/documentation/swiftui/view/swipeactions(edge:allowsfullswipe:content:)
- https://developer.android.com/develop/ui/compose/touch-input/pointer-input/migrate-swipeable

When that happens, we won't need a separate Composable for SwipeBoxes, as everything should be swipeable with a Modifier.

Reply:
>Besides, pairs are unstable, and this code will recompose on every single pixel of swipe gesture without strong skipping mode enabled

I've verified that my latest commit only recomposes the SwipeBox when the State is changed to fully expanded, other recompositions are skipped.

![image](https://github.com/user-attachments/assets/9b1019f0-4401-40b4-b843-7d5c39b06899)

Marking this as resolved, as I think the API is cleaner, and the main concerns have been addressed.